server/EXCEPTIONS.md

```diff
@@ -0,0 +1,168 @@
+# Exception handling in our backend
+
+## Quickstart
+
+To generate errors in the backend that will be passed to the
+frontend, throw a [Nest-defined HttpException](https://docs.nestjs.com/exception-filters#throwing-standard-exceptions), <u>_generally within a service_</u>:
+
+```js
+  throw new HttpException(responseBody, status)
+```
+
+Where...
+  - `status` is one of `HttpStatus` enum codes. E.g. `HttpStatus.UNAUTHORIZED`
+  - `responseBody` argument is an object with the following schema.
+    ```js
+       const responseBody = {
+        code: 'APPLICANT_PORTAL_ERROR_CODE',
+        message: 'Applicant Portal could not find coffee. CRM: CRM ran out of tea.',
+      }
+    ```
+    - The `code` property is developer-defined, and should be a unique, concise,
+    descriptive identifier for the type of error. Users can provide this code to
+    help developers locate the code that yielded the error.
+    - The `message` property is a short, human-readable sentence firstly
+    describing the error for the benefit of end users. Secondly, it should contain
+    errors relayed from CRM, if available.
+
+To scope the error to a specific part of the code, use try-catch blocks. For example,
+the following try-catch scopes the thrown error to an attempt to find a project.
+
+```js
+// services/my-project.service.ts
+getProject(){
+  try {
+    return myProject = this.get('dcp_project', ...);
+  } catch (e) {
+    throw new HttpException({
+      code: 'GET_PROJECTS',
+      message: `Failed to find a project for given id. CRM: ${e}`,
+    });
+  }
+}
+```
+
+Generally, when generating errors as described above, <u>*do so within Nest services*</u>.
+
+## Relaying Errors
+
+In the controller that consumes a service, _relay_ the error from the service by simply throwing the error as it bubbles up:
+
+```js
+// controllers/my-project.controller.ts
+@Get('/')
+  async getProject(@Session() session) {
+    try {
+      return this.myProjectService.getProject();
+    } catch (e) {
+      throw e; // simply relay the bubbling error
+    }
+  }
+}
+```
+
+Note how it is not necessary to construct a new HttpException in the controller to relay the error from the service. Instead, we simply use `throw` to pass along the error `e` from the service.
+
+Errors will *bubble* through the function call stack. The `catch` block in a try-catch will also catch errors generated from *within* functions that are called in the `try` block.
+
+Only the HttpException caught and thrown closest to or on the controller
+endpoint will be returned for a given request. To prevent confusion over which
```

Comment:
So does this mean that if a controller employs TWO services and both services error out that only one service's error will be shown?

Reply:
I actually realize I need to rewrite this part because of a nuance I missed. Standby..

Reply:
Have to handle things a little different with async/await. Reading up on it now.

Reply:
So essentially we need to make sure we also use `await` when we directly return async functions. This article was wonderful shedding light on this: https://itnext.io/error-handling-with-async-await-in-js-26c3f20bc06a

I also need to clarify the nuance that that any outer try-catch will only catch _uncaught_  errors. If a nested try-catch wants to relay an error (let an outer try-catch to handle something), it has to re-throw the error or generate a new error.
Updating the README now. I'll also try to update it to address your original question and your other comments. Thanks for these!

Reply:
Re your question:
If a controller employs two services, and both error out, I think only the first one that errors out will be shown. I think code execution stops when the first error is hit